src/v/raft/state_machine_manager.cc

```diff
@@ -431,7 +431,9 @@ ss::future<> state_machine_manager::background_apply_fiber(
   entry_ptr entry, ssx::semaphore_units units) {
     while (!_as.abort_requested() && entry->stm->next() < _next) {
         storage::log_reader_config config(
-          entry->stm->next(), _next, ss::default_priority_class());
+          entry->stm->next(),
```

Comment:
> This lead to a situation in which there may be no stms suitable to be handled in main apply fiber

how is this possible? In the situation you described (fe)

raft_committed = 105
stm_mgr._next = 100
bg_stm._next = 101

bg apply fiber exits (101 < 100 = false) .. main apply loop tries to read [100, 105] to catchup , no? what exactly is stalling it.

Reply:
in this case the `bg_stm` is not eligible for main loop as its `_next != stm_manager::_next`

Reply:
I still don't fully get it :-D 

> in this case the bg_stm is not eligible for main loop as its _next != stm_manager::_next

right.. but stm_manager::_next should advance right? (in the next iteration of foreground_apply) making all stms eligible in the apply after that. If I'm to guess stm_manager::_next is stalled for some reason preventing it to catch up, but unclear to me why.

Reply:
chatted offline.. so this scenario is possible when _all_ stms run into this "off by one" condition at the same time , so in the next iteration of apply, no stm is eligible for foreground apply and the background fibers have already exited, so effectively resulting in a stall 🛑